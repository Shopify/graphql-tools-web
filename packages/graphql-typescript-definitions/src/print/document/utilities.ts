import {ucFirst} from 'change-case';
import {GraphQLCompositeType} from 'graphql';
import {Field} from 'graphql-tool-utilities/ast';

export class ObjectStack {
  private seenFields = new Set<string>();

  get name(): string {
    const {parent, field, isFragment, type} = this;
    const fieldName = field ? ucFirst(field.responseName) : '';
    const name = `${parent ? parent.name : ''}${fieldName}`;
    return isFragment ? `${name}${type ? type.name : 'Other'}` : name;
  }

  constructor(
    private type?: GraphQLCompositeType,
    private field?: Field,
    private parent?: ObjectStack,
    private isFragment = false,
  ) {}

  nested(field: Field, type: GraphQLCompositeType) {
    return new ObjectStack(type, field, this);
  }

  fragment(type?: GraphQLCompositeType) {
    return new ObjectStack(type, this.field, this.parent, true);
  }

  sawField(field: Field) {
    this.seenFields.add(field.responseName);
  }

  hasSeenField(field: Field) {
    return this.seenFields.has(field.responseName);
  }
}
