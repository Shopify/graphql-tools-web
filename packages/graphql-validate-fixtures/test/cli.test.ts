import {resolve} from 'path';
import {execSync} from 'child_process';
import {stripFullFilePaths} from '../../../test/utilities';

const scriptPath = resolve(__dirname, '../bin/graphql-validate-fixtures');
const rootFixturePath = resolve(__dirname, 'fixtures');

describe('cli', () => {
  it('succeeds when there are no fixture errors', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('all-clear')),
    ).not.toThrowError();
  });

  it('succeeds for a multi-project when there are no fixture errors', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('multi-project')),
    ).not.toThrowError();
  });

  it('fails when there are ambiguous fixture names', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('ambiguous-operation')),
    ).toThrowErrorMatchingSnapshot();
  });

  it('fails when a fixture matches no operations', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('missing-operation')),
    ).toThrowErrorMatchingSnapshot();
  });

  it('fails when there are fixture errors', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('fixture-errors')),
    ).toThrowErrorMatchingSnapshot();
  });

  it('fails when there are syntax errors', () => {
    expect(() =>
      exec(cliCommandForFixtureDirectory('fixture-invalid')),
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      exec(cliCommandForFixtureDirectory('malformed-query')),
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      exec(cliCommandForFixtureDirectory('missing-schema')),
    ).toThrowErrorMatchingSnapshot();
  });

  it('hides passing fixtures by default', () => {
    expect(
      execSync(cliCommandForFixtureDirectory('all-clear')).toString(),
    ).toMatchSnapshot();
  });

  it('shows passes fixtures when the show-passes flag is true', () => {
    const showPasses = true;
    expect(
      execSync(
        cliCommandForFixtureDirectory('all-clear', showPasses),
      ).toString(),
    ).toMatchSnapshot();
  });
});

function exec(command: string) {
  try {
    const result = execSync(command);
    return stripFullFilePaths(result);
  } catch (error) {
    error.message = stripFullFilePaths(error.message);
    throw error;
  }
}

function cliCommandForFixtureDirectory(fixture: string, showPasses = false) {
  const fixtureDirectory = resolve(rootFixturePath, fixture);
  return [
    scriptPath,
    `'${resolve(fixtureDirectory, 'fixtures/**/*.json')}'`,
    `--cwd '${fixtureDirectory}'`,
    showPasses ? `--show-passes` : '',
  ]
    .join(' ')
    .trim();
}
