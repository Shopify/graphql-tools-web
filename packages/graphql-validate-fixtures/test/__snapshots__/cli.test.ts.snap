// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli fails when a fixture matches no operations 1`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/missing-operation/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/missing-operation'"`;

exports[`cli fails when there are ambiguous fixture names 1`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/ambiguous-operation/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/ambiguous-operation'"`;

exports[`cli fails when there are fixture errors 1`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/fixture-errors/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/fixture-errors'"`;

exports[`cli fails when there are syntax errors 1`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/fixture-invalid/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/fixture-invalid'"`;

exports[`cli fails when there are syntax errors 2`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/malformed-query/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/malformed-query'"`;

exports[`cli fails when there are syntax errors 3`] = `"Command failed: packages/graphql-validate-fixtures/bin/graphql-validate-fixtures 'packages/graphql-validate-fixtures/test/fixtures/missing-schema/fixtures/**/*.json' --cwd 'packages/graphql-validate-fixtures/test/fixtures/missing-schema'"`;

exports[`cli hides passing fixtures by default 1`] = `
"

3 passed
"
`;

exports[`cli shows passes fixtures when the show-passes flag is true 1`] = `
"
 PASS  fixtures/another-another-fixture.json
 PASS  fixtures/another-fixture.json
 PASS  fixtures/MyQuery/fixture.json

3 passed
"
`;
